File: frontend/pages/email-accounts.js
import { useEffect, useState } from 'react';
import { useRouter } from 'next/router';
import Layout from '../components/Layout';
import { api } from '../lib/api';
import {
  Button, Confirm, Empty, Field, Icons, Input, Modal, Panel, Select, Spinner, StatusBadge, fmtDate, useToast,
} from '../components/ui';

const EMPTY_FORM = {
  provider: 'smtp', email: '', display_name: '', app_password: '',
  smtp_host: 'smtp.gmail.com', smtp_port: 465, smtp_secure: true, smtp_username: '',
};

export default function EmailAccounts() {
  const toast = useToast();
  const router = useRouter();
  const [accounts, setAccounts] = useState(null);
  const [smtpOpen, setSmtpOpen] = useState(false);
  const [editAccount, setEditAccount] = useState(null);
  const [form, setForm] = useState(EMPTY_FORM);
  const [connectBusy, setConnectBusy] = useState(false);
  const [confirmDisconnect, setConfirmDisconnect] = useState(null);
  const [confirmDelete, setConfirmDelete] = useState(null);
  const [testingId, setTestingId] = useState(null);
  const [busyId, setBusyId] = useState(null);

  useEffect(() => {
    if (router.query.connected === '1') toast('Gmail account connected', 'success');
    if (router.query.oauth_error === '1') toast('OAuth connection failed or was cancelled', 'error');
    router.replace('/email-accounts', undefined, { shallow: true });
  }, [router.query]);

  const load = () => {
    api('/api/email-accounts').then(setAccounts).catch((e) => toast(e.message, 'error'));
  };
  useEffect(() => { load(); }, []);

  const connectGoogle = async () => {
    setConnectBusy(true);
    try {
      const res = await api('/api/email-accounts/connect');
      window.location.href = res.authorize_url;
    } catch (e) {
      toast(e.message, 'error');
      setConnectBusy(false);
    }
  };

  const addSmtp = async () => {
    try {
      await api('/api/email-accounts', { method: 'POST', body: form });
      setSmtpOpen(false);
      setForm(EMPTY_FORM);
      toast('SMTP account added', 'success');
      load();
    } catch (e) {
      toast(e.message, 'error');
    }
  };

  const saveEdit = async () => {
    try {
      const body = { ...editAccount };
      delete body.id; delete body.status; delete body.created_at; delete body.is_default;
      await api(`/api/email-accounts/${editAccount.id}`, { method: 'PATCH', body });
      toast('Account updated', 'success');
      setEditAccount(null);
      load();
    } catch (e) {
      toast(e.message, 'error');
    }
  };

  const doDisconnect = async () => {
    try {
      await api(`/api/email-accounts/${confirmDisconnect.id}/disconnect`, { method: 'POST' });
      toast('Account disconnected', 'success');
      setConfirmDisconnect(null);
      load();
    } catch (e) {
      toast(e.message, 'error');
    }
  };

  const doDelete = async () => {
    try {
      await api(`/api/email-accounts/${confirmDelete.id}`, { method: 'DELETE' });
      toast('Account deleted', 'success');
      setConfirmDelete(null);
      load();
    } catch (e) {
      toast(e.message, 'error');
    }
  };

  const setDefault = async (a) => {
    setBusyId(a.id);
    try {
      await api(`/api/email-accounts/${a.id}/default`, { method: 'POST' });
      toast(`Default sending account set to ${a.email}`, 'success');
      load();
    } catch (e) {
      toast(e.message, 'error');
    } finally {
      setBusyId(null);
    }
  };

  const testConnection = async (a) => {
    setTestingId(a.id);
    try {
      const res = await api(`/api/email-accounts/${a.id}/test`, { method: 'POST' });
      toast(res.message || 'Connection verified', 'success');
      load();
    } catch (e) {
      toast(e.message, 'error');
      load();
    } finally {
      setTestingId(null);
    }
  };

  return (
    <Layout title="Email Accounts" breadcrumb={<span>Email Accounts</span>}>
      <div className="page-head">
        <h1>Email Accounts</h1>
        <div className="muted">
          Connect your Gmail or SMTP accounts to send emails from your own address. Gmail uses Google's
          OAuth flow — we never ask for or store your Gmail password.
        </div>
      </div>

      <div className="toolbar">
        <Button disabled={connectBusy} onClick={connectGoogle}>
          {connectBusy ? <Spinner /> : Icons.google} Connect Google / Gmail
        </Button>
        <Button variant="secondary" onClick={() => setSmtpOpen(true)}>
          {Icons.plus} Add SMTP
        </Button>
      </div>

      <div className="grid" style={{ gridTemplateColumns: 'repeat(auto-fit, minmax(360px, 1fr))' }}>
        {accounts === null ? (
          <Spinner />
        ) : accounts.length === 0 ? (
          <Panel>
            <Empty message="No email accounts connected yet. Connect your Gmail to start sending." />
          </Panel>
        ) : (
          accounts.map((a) => (
            <Panel key={a.id} title={
              <span className="flex">
                {a.provider === 'google' ? Icons.google : Icons.email} {a.email}
                {a.is_default && <StatusBadge status="Default" tone="blue" />}
              </span>
            }
              actions={
                a.status === 'connected' ? (
                  <StatusBadge status="Connected" tone="green" />
                ) : (
                  <StatusBadge status={a.status} tone={a.status === 'disconnected' ? 'gray' : 'red'} />
                )
              }>
              <div className="field">
                <div className="label">Provider</div>
                <span className="muted">
                  {a.provider === 'google' ? 'Google (OAuth)' : 'SMTP'}
                  {a.provider === 'smtp' && ` · ${a.smtp_host}:${a.smtp_port}${a.smtp_secure ? ' (SSL/TLS)' : ''}`}
                </span>
              </div>
              {a.display_name && (
                <div className="field">
                  <div className="label">Display name</div>
                  <span className="muted">{a.display_name}</span>
                </div>
              )}
              <div className="field">
                <div className="label">Added</div>
                <span className="muted">{fmtDate(a.created_at)}</span>
              </div>
              <div className="flex" style={{ gap: 8, flexWrap: 'wrap' }}>
                <Button variant="secondary" sm disabled={testingId === a.id} onClick={() => testConnection(a)}>
                  {testingId === a.id ? <Spinner /> : Icons.check} Test
                </Button>
                {!a.is_default && (
                  <Button variant="secondary" sm disabled={busyId === a.id} onClick={() => setDefault(a)}>
                    {busyId === a.id ? <Spinner /> : Icons.check} Set as default
                  </Button>
                )}
                {a.provider === 'smtp' && (
                  <Button variant="secondary" sm onClick={() => setEditAccount(a)}>{Icons.edit} Edit</Button>
                )}
                {a.status === 'connected' && (
                  <Button variant="secondary" sm onClick={() => setConfirmDisconnect(a)}>
                    {Icons.x} Disconnect
                  </Button>
                )}
                <Button variant="danger" sm onClick={() => setConfirmDelete(a)}>
                  {Icons.trash} Delete
                </Button>
              </div>
            </Panel>
          ))
        )}
      </div>

      <Modal open={smtpOpen} title="Add SMTP account" onClose={() => setSmtpOpen(false)}
        footer={
          <>
            <Button variant="secondary" onClick={() => setSmtpOpen(false)}>Cancel</Button>
            <Button onClick={addSmtp}>Connect</Button>
          </>
        }>
        <p className="muted mb-16" style={{ marginTop: -6 }}>
          For Gmail use an <b>app password</b> (Settings → Security → App passwords). Never your regular password.
        </p>
        <Field label="Email">
          <Input type="email" required value={form.email} onChange={(e) => setForm({ ...form, email: e.target.value })} placeholder="you@example.com" />
        </Field>
        <Field label="Display name (optional)">
          <Input value={form.display_name} onChange={(e) => setForm({ ...form, display_name: e.target.value })} placeholder="Your Name" />
        </Field>
        <div className="grid" style={{ gridTemplateColumns: '1fr 1fr', gap: 12 }}>
          <Field label="SMTP host" help="Default: smtp.gmail.com">
            <Input value={form.smtp_host} onChange={(e) => setForm({ ...form, smtp_host: e.target.value })} placeholder="smtp.gmail.com" />
          </Field>
          <Field label="Port">
            <Input type="number" value={form.smtp_port} onChange={(e) => setForm({ ...form, smtp_port: Number(e.target.value) })} placeholder="465" />
          </Field>
        </div>
        <Field label="Connection">
          <Select value={form.smtp_secure ? 'ssl' : 'tls'} onChange={(e) => setForm({ ...form, smtp_secure: e.target.value === 'ssl' })}>
            <option value="ssl">SSL/TLS (implicit, port 465)</option>
            <option value="tls">STARTTLS (port 587)</option>
          </Select>
        </Field>
        <Field label="SMTP username (optional)" help="Defaults to the email above.">
          <Input value={form.smtp_username} onChange={(e) => setForm({ ...form, smtp_username: e.target.value })} placeholder={form.email || 'you@example.com'} />
        </Field>
        <Field label="Password / app password" help="Stored encrypted on the server.">
          <Input type="password" required value={form.app_password} onChange={(e) => setForm({ ...form, app_password: e.target.value })} placeholder="password" />
        </Field>
      </Modal>

      <Modal open={!!editAccount} title="Edit SMTP account" onClose={() => setEditAccount(null)}
        footer={
          <>
            <Button variant="secondary" onClick={() => setEditAccount(null)}>Cancel</Button>
            <Button onClick={saveEdit}>Save</Button>
          </>
        }>
        <Field label="Email">
          <Input type="email" required value={editAccount?.email || ''} onChange={(e) => setEditAccount({ ...editAccount, email: e.target.value })} />
        </Field>
        <Field label="Display name (optional)">
          <Input value={editAccount?.display_name || ''} onChange={(e) => setEditAccount({ ...editAccount, display_name: e.target.value })} />
        </Field>
        <div className="grid" style={{ gridTemplateColumns: '1fr 1fr', gap: 12 }}>
          <Field label="SMTP host">
            <Input value={editAccount?.smtp_host || ''} onChange={(e) => setEditAccount({ ...editAccount, smtp_host: e.target.value })} />
          </Field>
          <Field label="Port">
            <Input type="number" value={editAccount?.smtp_port || ''} onChange={(e) => setEditAccount({ ...editAccount, smtp_port: Number(e.target.value) })} />
          </Field>
        </div>
        <Field label="Connection">
          <Select value={editAccount?.smtp_secure ? 'ssl' : 'tls'} onChange={(e) => setEditAccount({ ...editAccount, smtp_secure: e.target.value === 'ssl' })}>
            <option value="ssl">SSL/TLS (implicit, port 465)</option>
            <option value="tls">STARTTLS (port 587)</option>
          </Select>
        </Field>
        <Field label="SMTP username (optional)">
          <Input value={editAccount?.smtp_username || ''} onChange={(e) => setEditAccount({ ...editAccount, smtp_username: e.target.value })} />
        </Field>
        <Field label="New password / app password (optional)" help="Leave blank to keep the current one. Stored encrypted.">
          <Input type="password" value={editAccount?.app_password || ''} onChange={(e) => setEditAccount({ ...editAccount, app_password: e.target.value })} />
        </Field>
      </Modal>

      <Confirm open={!!confirmDisconnect} title="Disconnect account" danger
        message={`Disconnect ${confirmDisconnect?.email}? Its credentials will be removed. You can reconnect later.`}
        confirmLabel="Disconnect" onCancel={() => setConfirmDisconnect(null)} onConfirm={doDisconnect} />

      <Confirm open={!!confirmDelete} title="Delete account" danger
        message={`Delete ${confirmDelete?.email}? This removes the account permanently. Emails already sent stay in your history.`}
        confirmLabel="Delete" onCancel={() => setConfirmDelete(null)} onConfirm={doDelete} />
    </Layout>
  );
}

File: backend/routers/email_accounts.py
import smtplib

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from .. import gmail
from ..config import API_BASE, FRONTEND_URL
from ..database import get_db
from ..encryption import decrypt_plaintext, encrypt_plaintext
from ..models import EmailAccount, User
from ..schemas import EmailAccountIn, EmailAccountOut, EmailAccountUpdate
from ..security import create_access_token, decode_access_token, get_current_user

router = APIRouter(prefix='/api/email-accounts', tags=['email accounts'])


def _redirect_uri():
    return f'{API_BASE}/api/email-accounts/callback'


def _out(account):
    return EmailAccountOut(
        id=account.id,
        provider=account.provider,
        email=account.email,
        display_name=account.display_name or '',
        smtp_host=account.smtp_host or '',
        smtp_port=account.smtp_port or 0,
        smtp_secure=account.smtp_secure if account.smtp_secure is not None else True,
        smtp_username=account.smtp_username or '',
        is_default=bool(account.is_default),
        status=account.status,
        created_at=account.created_at.isoformat() if account.created_at else None,
    )


def _load(db, user, account_id):
    account = (
        db.query(EmailAccount)
        .filter(EmailAccount.id == account_id, EmailAccount.user_id == user.id)
        .first()
    )
    if not account:
        raise HTTPException(status_code=404, detail='Email account not found')
    return account


def _set_default(db, user, account, is_default):
    if is_default:
        db.query(EmailAccount).filter(EmailAccount.user_id == user.id).update(
            {'is_default': False}, synchronize_session=False
        )
    account.is_default = bool(is_default)


@router.get('', response_model=list[EmailAccountOut])
def list_accounts(user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    accounts = (
        db.query(EmailAccount).filter(EmailAccount.user_id == user.id).order_by(EmailAccount.id).all()
    )
    return [_out(a) for a in accounts]


@router.post('', response_model=EmailAccountOut)
def add_smtp_account(
    payload: EmailAccountIn,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    if payload.provider == 'smtp' and not payload.app_password:
        raise HTTPException(status_code=400, detail='A password is required for SMTP accounts')
    email = payload.email.lower()
    existing = (
        db.query(EmailAccount)
        .filter(EmailAccount.user_id == user.id, EmailAccount.email == email)
        .first()
    )
    if existing:
        raise HTTPException(status_code=409, detail='This email account is already connected')
    count = db.query(EmailAccount).filter(EmailAccount.user_id == user.id).count()
    is_default = payload.is_default or count == 0
    account = EmailAccount(
        user_id=user.id,
        provider=payload.provider,
        email=email,
        display_name=payload.display_name,
        credentials_encrypted=encrypt_plaintext(payload.app_password) if payload.provider == 'smtp' else None,
        smtp_host=payload.smtp_host or 'smtp.gmail.com',
        smtp_port=payload.smtp_port or 465,
        smtp_secure=payload.smtp_secure if payload.smtp_secure is not None else True,
        smtp_username=payload.smtp_username,
        is_default=is_default,
        status='connected',
    )
    db.add(account)
    db.commit()
    db.refresh(account)
    return _out(account)


@router.get('/connect')
def connect_google(user: User = Depends(get_current_user)):
    if not gmail.GOOGLE_CLIENT_ID or not gmail.GOOGLE_CLIENT_SECRET:
        raise HTTPException(
            status_code=500,
            detail='Google OAuth is not configured on the server (GOOGLE_CLIENT_ID/GOOGLE_CLIENT_SECRET)',
        )
    state = create_access_token(user.id)
    return {'authorize_url': gmail.build_authorize_url(state, _redirect_uri())}


@router.get('/callback')
def oauth_callback(code: str, state: str, error: str = '', db: Session = Depends(get_db)):
    if error:
        return f'<script>window.location.href="{FRONTEND_URL}/email-accounts?oauth_error=1"</script>'
    user_id = decode_access_token(state)
    if not user_id:
        raise HTTPException(status_code=401, detail='Invalid OAuth state')
    try:
        tokens = gmail.exchange_code(code, _redirect_uri())
    except Exception as exc:
        return f'<script>window.location.href="{FRONTEND_URL}/email-accounts?oauth_error=1"</script>'
    email = (tokens.get('email') or '').lower()
    account = (
        db.query(EmailAccount)
        .filter(EmailAccount.user_id == user_id, EmailAccount.email == email)
        .first()
    )
    if not account:
        count = db.query(EmailAccount).filter(EmailAccount.user_id == user_id).count()
        account = EmailAccount(
            user_id=user_id,
            provider='google',
            email=email,
            is_default=count == 0,
            status='connected',
        )
        db.add(account)
    account.credentials_encrypted = encrypt_plaintext(tokens['refresh_token'])
    account.status = 'connected'
    db.commit()
    return f'<script>window.location.href="{FRONTEND_URL}/email-accounts?connected=1"</script>'


@router.patch('/{account_id}', response_model=EmailAccountOut)
def update_account(
    account_id: int,
    payload: EmailAccountUpdate,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    account = _load(db, user, account_id)
    if payload.app_password:
        account.credentials_encrypted = encrypt_plaintext(payload.app_password)
        account.status = 'connected'
    for field in ('email', 'display_name', 'smtp_host', 'smtp_port', 'smtp_secure', 'smtp_username'):
        value = getattr(payload, field)
        if value is not None:
            setattr(account, field, value)
    if payload.is_default is not None:
        _set_default(db, user, account, payload.is_default)
    db.commit()
    db.refresh(account)
    return _out(account)


@router.post('/{account_id}/default', response_model=EmailAccountOut)
def set_default(account_id: int, user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    account = _load(db, user, account_id)
    _set_default(db, user, account, True)
    db.commit()
    db.refresh(account)
    return _out(account)


@router.post('/{account_id}/disconnect')
def disconnect_account(account_id: int, user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    account = _load(db, user, account_id)
    account.status = 'disconnected'
    account.credentials_encrypted = None
    db.commit()
    return {'ok': True}


@router.delete('/{account_id}')
def delete_account(account_id: int, user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    account = _load(db, user, account_id)
    db.delete(account)
    db.commit()
    return {'ok': True}


@router.post('/{account_id}/test')
def test_account(account_id: int, user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    account = _load(db, user, account_id)
    if not account.credentials_encrypted:
        raise HTTPException(status_code=400, detail='No credentials stored for this account')
    try:
        if account.provider == 'google':
            refresh_token = decrypt_plaintext(account.credentials_encrypted)
            gmail.build_gmail_service(refresh_token)
        else:
            password = decrypt_plaintext(account.credentials_encrypted)
            host = account.smtp_host or 'smtp.gmail.com'
            port = account.smtp_port or 465
            secure = account.smtp_secure if account.smtp_secure is not None else True
            username = account.smtp_username or account.email
            if secure:
                with smtplib.SMTP_SSL(host, port) as smtp:
                    smtp.login(username, password)
            else:
                with smtplib.SMTP(host, port) as smtp:
                    smtp.starttls()
                    smtp.login(username, password)
        account.status = 'connected'
        db.commit()
        return {'ok': True, 'status': 'connected', 'message': f'Connection to {account.email} verified'}
    except Exception as exc:
        account.status = 'error'
        db.commit()
        raise HTTPException(status_code=400, detail=f'Connection failed: {exc}')


def get_refresh_token(account):
    if not account or not account.credentials_encrypted:
        return None
    return decrypt_plaintext(account.credentials_encrypted)